src/agent/application.cpp

```diff
@@ -91,29 +91,45 @@ Application::Application(const std::string               &aInterfaceName,
 #if OTBR_ENABLE_VENDOR_SERVER
     , mVendorServer(vendor::VendorServer::newInstance(*this))
 #endif
+    , mThreadEnabled(false)
 {
     OTBR_UNUSED_VARIABLE(aRestListenAddress);
     OTBR_UNUSED_VARIABLE(aRestListenPort);
 }
 
-void Application::Init(void)
+void Application::OnThreadEnabledChanged(bool aThreadEnabled)
 {
-    mNcp.Init();
+    VerifyOrExit(mThreadEnabled != aThreadEnabled);
 
-#if OTBR_ENABLE_MDNS
-    mPublisher->Start();
+    mThreadEnabled = aThreadEnabled;
+
+#if OTBR_ENABLE_BACKBONE_ROUTER
+    mBackboneAgent.SetEnabled(mThreadEnabled);
 #endif
+
 #if OTBR_ENABLE_BORDER_AGENT
-    mBorderAgent.SetEnabled(true);
-#endif
-#if OTBR_ENABLE_BACKBONE_ROUTER
-    mBackboneAgent.Init();
+    mBorderAgent.SetEnabled(mThreadEnabled);
 #endif
 #if OTBR_ENABLE_SRP_ADVERTISING_PROXY
-    mAdvertisingProxy.SetEnabled(true);
+    mAdvertisingProxy.SetEnabled(mThreadEnabled);
 #endif
 #if OTBR_ENABLE_DNSSD_DISCOVERY_PROXY
-    mDiscoveryProxy.SetEnabled(true);
+    mDiscoveryProxy.SetEnabled(mThreadEnabled);
```

Comment:
Is it safe/okay to possibly delay the call to `SetEnabled()` for `mAdvertisingProxy`, `mDiscoveryProxy` and `mBackboneAgent`?

Can we have the OT core stack trying to interact with these platform modules and cause issue if they are not enabled?

Reply:
Thinking more on this: how about we limit this only for `mBorderAgent.SetEnabled()` (which I think the main use-case here)? And still enable other components from `Init()`?